File: mbtsdk/src/main/java/command/DeviceCommandEvent.java
package command;

import java.nio.ByteBuffer;
import java.util.HashMap;

/**
 * A class that contains all currently used device commands (mailbox & other commands) codes.
 * One code has one specific functionality.
 */
public enum DeviceCommandEvent {

    /**
     * Event codes related to OAD Events
     */
    MBX_START_OTA_TXF((byte)0x03),           // Used by appli to request an OTA update (provides software major and minor in payload)
    MBX_OTA_IDX_RESET_EVT((byte)0x06),       // Notifies appli that we request a packet Idx reset
    MBX_OTA_STATUS_EVT((byte)0x07),   // Notifies appli with the status of the OTA transfer.
    OTA_STATUS_TRANSFER((byte)-1),   // Notifies SDK when an OAD packet has been transferred.
    OTA_BLUETOOTH_RESET((byte)-2),   // Notifies SDK when an the Bluetooth has been reset

    MBX_OTA_MODE_EVT((byte)0x05,
        new HashMap<String, Byte>() {{
            put(CMD_CODE_OTA_MODE_EVT_FAILED, ((byte)0x00));
            put(CMD_CODE_OTA_MODE_EVT_SUCCESS, ((byte)0xFF));
        }}
    ),

    /**
     * Event codes related to Device System Events
     */
    MBX_SET_ADS_CONFIG((byte)0x00),
    MBX_SET_AUDIO_CONFIG((byte)0x01),

    /**
     * Event codes related to Device Configuration Events
     */
    MBX_SET_PRODUCT_NAME((byte)0x02), // Product name configuration request
    MBX_SET_NOTCH_FILT((byte)0x0B),      // allows to hotswap the filters' parameters
    MBX_SET_BANDPASS_FILT((byte)0x0C),   // Set the signal bandwidth by changing the embedded bandpass filter
    MBX_SET_AMP_GAIN((byte)0x0D),        // Set the eeg signal amplifier gain
    MBX_P300_ENABLE((byte)0x0F),         // Enable or disable the p300 functionnality of the melomind.
    MBX_DC_OFFSET_ENABLE((byte)0x10),         // Enable or disable the DC offset measurement computation and sending.

    MBX_SET_SERIAL_NUMBER((byte)0x0A,
            (byte) 0x53, (byte) 0x4D),

    MBX_SET_EXTERNAL_NAME(MBX_SET_SERIAL_NUMBER.identifierCode,
            (byte) 0xAB, (byte) 0x21),

    /**
     * Event codes related to a reading operation
     */
    MBX_SYS_GET_STATUS((byte)0x08),   // allows to retrieve to system global status
    MBX_GET_EEG_CONFIG((byte)0x0E),      // Get the current configuration of the Notch filter, the bandpass filter, and the amplifier gain.
    GET_EEG_CONFIG_ADDITIONAL((byte) 0x00, (byte)0x00), // Additional info to get the current configuration .

    CMD_GET_BATTERY_VALUE ((byte)0x20),
    CMD_START_EEG_ACQUISITION((byte)0x24),
    CMD_STOP_EEG_ACQUISITION((byte)0x25),
    CMD_GET_DEVICE_CONFIG ((byte)0x50),
    CMD_GET_PSM_CONFIG((byte)0x51),
    CMD_GET_IMS_CONFIG((byte)0x52),
    CMD_SET_DEVICE_CONFIG((byte)0x53),
    CMD_GET_DEVICE_INFO((byte)0x54),

    START_FRAME((byte) 0x3C),
    PAYLOAD_LENGTH( (byte)0x00, (byte)0x00 ),
    COMPRESS((byte) 0x00),
    PACKET_ID((byte) 0x00 ),
    PAYLOAD((byte) 0x00), // Additional info to get the current configuration .

    /**
     * Event codes related to Audio Connection Events
     */
    MBX_CONNECT_IN_A2DP((byte)0x11,
            new byte[]{(byte) 0x25,(byte) 0xA2},
            new HashMap<String, Byte>() {{
                put(CMD_CODE_CONNECT_IN_A2DP_IN_PROGRESS,(byte)0x01);
                put(CMD_CODE_CONNECT_IN_A2DP_FAILED_BAD_BDADDR,(byte)0x02);
                put(CMD_CODE_CONNECT_IN_A2DP_FAILED_ALREADY_CONNECTED,(byte)0x04);
                put(CMD_CODE_CONNECT_IN_A2DP_FAILED_TIMEOUT,(byte)0x08);
                put(CMD_CODE_CONNECT_IN_A2DP_LINKKEY_INVALID,(byte)0x10);
                put(CMD_CODE_CONNECT_IN_A2DP_JACK_CONNECTED,(byte)0x20);
                put(CMD_CODE_CONNECT_IN_A2DP_SUCCESS,(byte)0x80);
                }
            }
        ),

    /**
     * Event codes related to Audio Reconnection Events
     */
    MBX_SYS_REBOOT_EVT((byte)0x09,
            (byte) 0x29,(byte) 0x08),

    /**
     * Event codes related to Audio Disconnection Events
     */
    MBX_DISCONNECT_IN_A2DP((byte)0x12,
        new byte[]{(byte) 0x85, (byte) 0x11},
        new HashMap<String, Byte>() {{
                put(CMD_CODE_DISCONNECT_IN_A2DP_FAILED, (byte) 0x01);
                put(CMD_CODE_DISCONNECT_IN_A2DP_SUCCESS, (byte) 0xFF);
            }
        }
    ),

    /**
     * Unused codes
     */
    MBX_LEAD_OFF_EVT((byte)0x04),            // Notifies app of a lead off modification
    MBX_BAD_EVT((byte)0xFF);

    public static final String CMD_CODE_CONNECT_IN_A2DP_IN_PROGRESS = "CMD_CODE_CONNECT_IN_A2DP_IN_PROGRESS";
    public static final String CMD_CODE_CONNECT_IN_A2DP_FAILED_BAD_BDADDR = "CMD_CODE_CONNECT_IN_A2DP_FAILED_BAD_BDADDR";
    public static final String CMD_CODE_CONNECT_IN_A2DP_FAILED_ALREADY_CONNECTED = "CMD_CODE_CONNECT_IN_A2DP_FAILED_ALREADY_CONNECTED";
    public static final String CMD_CODE_CONNECT_IN_A2DP_FAILED_TIMEOUT = "CMD_CODE_CONNECT_IN_A2DP_FAILED_TIMEOUT";
    public static final String CMD_CODE_CONNECT_IN_A2DP_LINKKEY_INVALID = "CMD_CODE_CONNECT_IN_A2DP_LINKKEY_INVALID";
    public static final String CMD_CODE_CONNECT_IN_A2DP_JACK_CONNECTED = "CMD_CODE_CONNECT_IN_A2DP_JACK_CONNECTED";
    public static final String CMD_CODE_CONNECT_IN_A2DP_SUCCESS = "CMD_CODE_CONNECT_IN_A2DP_SUCCESS";

    public static final String CMD_CODE_DISCONNECT_IN_A2DP_FAILED = "CMD_CODE_DISCONNECT_IN_A2DP_FAILED";
    public static final String CMD_CODE_DISCONNECT_IN_A2DP_SUCCESS = "CMD_CODE_DISCONNECT_IN_A2DP_SUCCESS";

    public static final String CMD_CODE_OTA_MODE_EVT_FAILED = "CMD_CODE_OTA_MODE_EVT_FAILED";
    public static final String CMD_CODE_OTA_MODE_EVT_SUCCESS = "CMD_CODE_OTA_MODE_EVT_SUCCESS";

    /**
     * Unique identifier of the event.
     */
    private byte identifierCode;

    /**
     * Optional additional code associated to the event
     * to add security and avoid requests sent by hackers.
     */
    private byte[] additionalCodes;

    /**
     * Optional additional codes associated to the event
     * that holds the possible responses returned once the device command has been sent.
     */
    private HashMap<String,Byte> responseCodesMap;

    DeviceCommandEvent(byte identifierCode, byte[] additionalCodes, HashMap<String, Byte> responseCodesMap) {
        this.identifierCode = identifierCode;
        this.additionalCodes = additionalCodes;
        this.responseCodesMap = responseCodesMap;
    }

    DeviceCommandEvent(byte identifierCode, HashMap<String, Byte> responseCodesMap) {
        this.identifierCode = identifierCode;
        this.responseCodesMap = responseCodesMap;
    }

    DeviceCommandEvent(byte identifierCode, byte... additionalCodes) {
        this.identifierCode = identifierCode;
        this.additionalCodes = additionalCodes;
    }


    DeviceCommandEvent(byte identifierCode) {
        this.identifierCode = identifierCode;
    }

    /**
     * Returns the unique identifier of the command
     * @return the unique identifier of the command
     */
    public byte getIdentifierCode() {
        return identifierCode;
    }

    public byte[] getAssembledCodes() {
        return assembleCodes(new byte[]{identifierCode}, additionalCodes);
    }

    /**
     * Returns the optional additional code associated to the event
     * to add security and avoid requests sent by hackers.
     */
    public byte[] getAdditionalCodes() {
        return additionalCodes;
    }

    /**
     * Returns the ptional additional codes associated to the event
     * that holds the possible responses returned once the device command has been sent.
     */
    public Byte getResponseCodeForKey(String key) {
        return (responseCodesMap.containsKey(key) ? responseCodesMap.get(key) : null);
    }

    /**
     * Returns the event that has an identifier that matchs the identifier code
     * @param identifierCode the identifier code
     * @return the event that has an identifier that matchs the identifier code
     */
    public static DeviceCommandEvent getEventFromIdentifierCode(byte identifierCode){
        for (DeviceCommandEvent event : DeviceCommandEvent.values()) {
            if(identifierCode == event.identifierCode)
                return event;
        }
        return null;
    }

    public static byte[] assembleCodes(byte[]... codes){
        int bufferLength = 0;
        for (byte[] code : codes) {
            if(code != null){
                for (byte codeByte : code) {
                    bufferLength++;
                }
            }
        }
        ByteBuffer buffer = ByteBuffer.allocate(bufferLength);
        for (byte[] code : codes) {
            if (code != null) {
                for (byte codeByte : code) {
                    buffer.put(codeByte);
                }
            }
        }
        return buffer.array();
    }
}

File: sdk-v3/src/main/java/com/mybraintech/sdk/core/bluetooth/devices/melomind/DeviceCommandEvent.kt
package com.mybraintech.sdk.core.bluetooth.devices.melomind

import com.mybraintech.sdk.core.decodeFromHex
import java.nio.ByteBuffer
import java.util.*

/**
 * A class that contains all currently used device commands (mailbox & other commands) codes.
 * One code has one specific functionality.
 */
enum class DeviceCommandEvent {
  /**
   * Event codes related to OAD Events
   */
  MBX_START_OTA_TXF(0x03.toByte()),  // Used by appli to request an OTA update (provides software major and minor in payload)
  MBX_OTA_IDX_RESET_EVT(0x06.toByte()),  // Notifies appli that we request a packet Idx reset
  MBX_OTA_STATUS_EVT(0x07.toByte()),  // Notifies appli with the status of the OTA transfer.
  OTA_STATUS_TRANSFER((-1).toByte()),  // Notifies SDK when an OAD packet has been transferred.
  OTA_BLUETOOTH_RESET((-2).toByte()),  // Notifies SDK when an the Bluetooth has been reset
  MBX_OTA_MODE_EVT(0x05.toByte(),
      object : HashMap<String, Byte>() {
        init {
          put("CMD_CODE_OTA_MODE_EVT_FAILED", 0x00.toByte())
          put("CMD_CODE_OTA_MODE_EVT_SUCCESS", 0xFF.toByte())
        }
      }
  ),

  /**
   * Event codes related to Device System Events
   */
  MBX_SET_ADS_CONFIG(0x00.toByte()), MBX_SET_AUDIO_CONFIG(0x01.toByte()),

  /**
   * Event codes related to Device Configuration Events
   */
  MBX_SET_PRODUCT_NAME(0x02.toByte()),  // Product name configuration request
  MBX_SET_NOTCH_FILT(0x0B.toByte()),  // allows to hotswap the filters' parameters
  MBX_SET_BANDPASS_FILT(0x0C.toByte()),  // Set the signal bandwidth by changing the embedded bandpass filter
  MBX_SET_AMP_GAIN(0x0D.toByte()),  // Set the eeg signal amplifier gain
  MBX_P300_ENABLE(0x0F.toByte()),  // Enable or disable the p300 functionnality of the melomind.
  MBX_DC_OFFSET_ENABLE(0x10.toByte()),  // Enable or disable the DC offset measurement computation and sending.
  MBX_SET_SERIAL_NUMBER(0x0A.toByte(),
      0x53.toByte(), 0x4D.toByte()),
  MBX_SET_EXTERNAL_NAME(MBX_SET_SERIAL_NUMBER.identifierCode,
      0xAB.toByte(), 0x21.toByte()),

  /**
   * Event codes related to a reading operation
   */
  MBX_SYS_GET_STATUS(0x08.toByte()),  // allows to retrieve to system global status
  MBX_GET_EEG_CONFIG(0x0E.toByte()),  // Get the current configuration of the Notch filter, the bandpass filter, and the amplifier gain.
  GET_EEG_CONFIG_ADDITIONAL(0x00.toByte(), 0x00.toByte()),  // Additional info to get the current configuration .
  CMD_GET_BATTERY_VALUE(0x20.toByte()),
  CMD_START_EEG_ACQUISITION(0x24.toByte()),
  CMD_STOP_EEG_ACQUISITION(0x25.toByte()),
  CMD_GET_DEVICE_CONFIG(0x50.toByte()),
  CMD_GET_PSM_CONFIG(0x51.toByte()),
  CMD_GET_IMS_CONFIG(0x52.toByte()),
  CMD_SET_DEVICE_CONFIG(0x53.toByte()),
  CMD_GET_DEVICE_INFO(0x54.toByte()),
  START_FRAME(0x3C.toByte()),
  PAYLOAD_LENGTH(0x00.toByte(), 0x00.toByte()),
  COMPRESS(0x00.toByte()),
  PACKET_ID(0x00.toByte()),
  PAYLOAD(0x00.toByte()),  // Additional info to get the current configuration .

  /**
   * Event codes related to Audio Connection Events
   */
  MBX_CONNECT_IN_A2DP(0x11.toByte(), byteArrayOf(0x25.toByte(), 0xA2.toByte()),
      object : HashMap<String, Byte>() {
        init {
          put("CMD_CODE_CONNECT_IN_A2DP_IN_PROGRESS", 0x01.toByte())
          put("CMD_CODE_CONNECT_IN_A2DP_FAILED_BAD_BDADDR", 0x02.toByte())
          put("CMD_CODE_CONNECT_IN_A2DP_FAILED_ALREADY_CONNECTED", 0x04.toByte())
          put("CMD_CODE_CONNECT_IN_A2DP_FAILED_TIMEOUT", 0x08.toByte())
          put("CMD_CODE_CONNECT_IN_A2DP_LINKKEY_INVALID", 0x10.toByte())
          put("CMD_CODE_CONNECT_IN_A2DP_JACK_CONNECTED", 0x20.toByte())
          put("CMD_CODE_CONNECT_IN_A2DP_SUCCESS", 0x80.toByte())
        }
      }
  ),

  /**
   * Event codes related to Audio Reconnection Events
   */
  MBX_SYS_REBOOT_EVT(0x09.toByte(),
      0x29.toByte(), 0x08.toByte()),

  /**
   * Event codes related to Audio Disconnection Events
   */
  MBX_DISCONNECT_IN_A2DP(0x12.toByte(), byteArrayOf(0x85.toByte(), 0x11.toByte()),
      object : HashMap<String, Byte>() {
        init {
          put("CMD_CODE_DISCONNECT_IN_A2DP_FAILED", 0x01.toByte())
          put("CMD_CODE_DISCONNECT_IN_A2DP_SUCCESS", 0xFF.toByte())
        }
      }
  ),

  /**
   * Unused codes
   */
  MBX_LEAD_OFF_EVT(0x04.toByte()),  // Notifies app of a lead off modification
  MBX_BAD_EVT(0xFF.toByte());

  /**
   * Returns the unique identifier of the command
   * @return the unique identifier of the command
   */
  /**
   * Unique identifier of the event.
   */
  var identifierCode: Byte
    private set

  /**
   * Returns the optional additional code associated to the event
   * to add security and avoid requests sent by hackers.
   */
  /**
   * Optional additional code associated to the event
   * to add security and avoid requests sent by hackers.
   */
  var additionalCodes: ByteArray? = null
    private set

  /**
   * Optional additional codes associated to the event
   * that holds the possible responses returned once the device command has been sent.
   */
  private var responseCodesMap: HashMap<String, Byte>? = null

  constructor(identifierCode: Byte, additionalCodes: ByteArray, responseCodesMap: HashMap<String, Byte>?) {
    this.identifierCode = identifierCode
    this.additionalCodes = additionalCodes
    this.responseCodesMap = responseCodesMap
  }

  constructor(identifierCode: Byte, responseCodesMap: HashMap<String, Byte>) {
    this.identifierCode = identifierCode
    this.responseCodesMap = responseCodesMap
  }

  constructor(identifierCode: Byte, vararg additionalCodes: Byte) {
    this.identifierCode = identifierCode
    this.additionalCodes = additionalCodes
  }

  constructor(identifierCode: Byte) {
    this.identifierCode = identifierCode
  }

  val assembledCodes: ByteArray
    get() = assembleCodes(byteArrayOf(identifierCode), additionalCodes)

  /**
   * Returns the ptional additional codes associated to the event
   * that holds the possible responses returned once the device command has been sent.
   */
  fun getResponseCodeForKey(key: String): Byte {
    if (responseCodesMap == null) {
      return 0.toByte()
    }
    return if (responseCodesMap!!.containsKey(key)) responseCodesMap!![key]?:0.toByte() else 0.toByte()
  }

  companion object {
    const val CMD_CODE_CONNECT_IN_A2DP_IN_PROGRESS = "CMD_CODE_CONNECT_IN_A2DP_IN_PROGRESS"
    const val CMD_CODE_CONNECT_IN_A2DP_FAILED_BAD_BDADDR = "CMD_CODE_CONNECT_IN_A2DP_FAILED_BAD_BDADDR"
    const val CMD_CODE_CONNECT_IN_A2DP_FAILED_ALREADY_CONNECTED = "CMD_CODE_CONNECT_IN_A2DP_FAILED_ALREADY_CONNECTED"
    const val CMD_CODE_CONNECT_IN_A2DP_FAILED_TIMEOUT = "CMD_CODE_CONNECT_IN_A2DP_FAILED_TIMEOUT"
    const val CMD_CODE_CONNECT_IN_A2DP_LINKKEY_INVALID = "CMD_CODE_CONNECT_IN_A2DP_LINKKEY_INVALID"
    const val CMD_CODE_CONNECT_IN_A2DP_JACK_CONNECTED = "CMD_CODE_CONNECT_IN_A2DP_JACK_CONNECTED"
    const val CMD_CODE_CONNECT_IN_A2DP_SUCCESS = "CMD_CODE_CONNECT_IN_A2DP_SUCCESS"
    const val CMD_CODE_DISCONNECT_IN_A2DP_FAILED = "CMD_CODE_DISCONNECT_IN_A2DP_FAILED"
    const val CMD_CODE_DISCONNECT_IN_A2DP_SUCCESS = "CMD_CODE_DISCONNECT_IN_A2DP_SUCCESS"
    const val CMD_CODE_OTA_MODE_EVT_FAILED = "CMD_CODE_OTA_MODE_EVT_FAILED"
    const val CMD_CODE_OTA_MODE_EVT_SUCCESS = "CMD_CODE_OTA_MODE_EVT_SUCCESS"

    /**
     * Returns the event that has an identifier that matchs the identifier code
     * @param identifierCode the identifier code
     * @return the event that has an identifier that matchs the identifier code
     */
    fun getEventFromIdentifierCode(identifierCode: Byte): DeviceCommandEvent? {
      for (event in values()) {
        if (identifierCode == event.identifierCode) return event
      }
      return null
    }

    @JvmStatic
    fun assembleCodes(vararg codes: ByteArray?): ByteArray {
      var bufferLength = 0
      for (code in codes) {
        if (code != null) {
          for (codeByte in code) {
            bufferLength++
          }
        }
      }
      val buffer = ByteBuffer.allocate(bufferLength)
      for (code in codes) {
        if (code != null) {
          for (codeByte in code) {
            buffer.put(codeByte)
          }
        }
      }
      return buffer.array()
    }
  }
}